src/cluster_resources.rs

```diff
@@ -0,0 +1,393 @@
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    pub async fn finalize(self, client: &Client) -> OperatorResult<()> {
+        self.delete_orphaned_resources::<Service>(client).await?;
+        self.delete_orphaned_resources::<StatefulSet>(client)
+            .await?;
+        self.delete_orphaned_resources::<DaemonSet>(client).await?;
+        self.delete_orphaned_resources::<ConfigMap>(client).await?;
+
+        Ok(())
+    }
+
+    /// Deletes all deployed resources which are labelled as if they belong to this cluster
+    /// instance but are not contained in the given list.
+    ///
+    /// If it is forbidden to list the resources of the given kind then it is assumed that the
+    /// caller is not in charge of these resources, the deletion is skipped, and no error is
+    /// returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    async fn delete_orphaned_resources<T: ClusterResource>(
+        &self,
+        client: &Client,
+    ) -> OperatorResult<()> {
+        match self.list_deployed_cluster_resources::<T>(client).await {
+            Ok(deployed_cluster_resources) => {
+                let orphaned_resources = deployed_cluster_resources
+                    .into_iter()
+                    .filter(|r| !self.resources.contains(&r.into()))
+                    .collect::<Vec<_>>();
+
+                if !orphaned_resources.is_empty() {
+                    info!(
+                        "Deleting orphaned {}: {}",
+                        T::plural(&()),
+                        ClusterResources::print_resources(&orphaned_resources),
+                    );
+                    for resource in orphaned_resources.iter() {
+                        client.delete(resource).await?;
+                    }
+                }
+
+                Ok(())
+            }
+            Err(Error::KubeError {
+                source: kube::Error::Api(ErrorResponse { code: 403, .. }),
+            }) => {
+                debug!(
+                    "Skipping deletion of orphaned {} because the operator is not allowed to list \
+                      them and is therefore probably not in charge of them.",
+                    T::plural(&())
+                );
+                Ok(())
+            }
+            Err(error) => Err(error),
+        }
+    }
+
+    /// Creates a string containing the names and if present namespaces of the given resources
+    /// sorted by name and separated with commas.
+    fn print_resources<T: ClusterResource>(resources: &[T]) -> String {
+        let mut output = resources
+            .iter()
+            .map(ClusterResources::print_resource)
+            .collect::<Vec<_>>();
+        output.sort();
+        output.join(", ")
+    }
+
+    /// Creates a string containing the name and if present namespace of the given resource.
+    fn print_resource<T: ClusterResource>(resource: &T) -> String {
```

Comment:
Should this maybe be 'pub'? Or, perhaps a debug impl. that just calls this fn ?

Reply:
> Should this maybe be 'pub'?

This is just a helper function. If its functionality is of public interest then it should be located in the implementation of a `Resource` or `ResourceExt` struct. I would not make arbitrary functions public because then they cannot be changed again without breaking the clients of this public API. Therefore I would keep the public interface as small as necessary to increase the evolvability.

> Or, perhaps a debug impl. that just calls this fn ?

Traits can only be implemented on trait objects. `Resource` cannot be made into an object because it contains associated functions without a `self` parameter, like `kind`. See https://doc.rust-lang.org/reference/items/traits.html#object-safety